admin_ui/src/modules/translations.ts

```diff
@@ -0,0 +1,84 @@
+import axios from "axios"
+
+import {
+    TranslationsListAPIResponse,
+    TranslationListItemAPI,
+    TranslationAPIResponse
+} from "@/interfaces"
+import i18n from "@/i18n"
+
+const BASE_URL = process.env.VUE_APP_BASE_URI
+
+const DEFAULT_LANGUAGE_KEY = "piccoloAdminDefaultLanguage"
+
+/**
+ * Stores the user's default language in localStorage. If the user selects a
+ * language, then this becomes their new default language, otherwise it's
+ * defined by the API.
+ */
+const localStorageUtils = {
+    getDefaultLanguage: (): string | null => {
+        return localStorage.getItem(DEFAULT_LANGUAGE_KEY)
+    },
+    setDefaultLanguage: (value: string) => {
+        return localStorage.setItem(DEFAULT_LANGUAGE_KEY, value)
+    }
+}
+
+export default {
+    state: {
+        defaultLanguage: null as string | null,
+        translations: null as string | TranslationListItemAPI[]
+    },
+    mutations: {
+        updateDefaultLanguage(state, value: string) {
+            state.defaultLanguage = value
+        },
+        updateTranslations(state, value: object) {
+            state.translations = value
+        }
+    },
+    actions: {
+        /**
+         * This fetches a list of all available translations (not the
+         * translations themselves).
+         */
+        async fetchTranslations(context) {
+            const response = await axios.get<TranslationsListAPIResponse>(
+                `${BASE_URL}translations/`
+            )
+
+            context.commit("updateTranslations", response.data.translations)
+
+            if (!context.state.defaultLanguage) {
+                context.commit(
+                    "updateDefaultLanguage",
+                    response.data.default_language_code
+                )
+            }
+        },
+        /**
+         * Fetch the translations for a certain language.
+         */
+        async fetchTranslation(context, languageCode: string) {
+            const response = await axios.get<TranslationAPIResponse>(
+                `${BASE_URL}translations/${languageCode}/`
+            )
+            i18n.setLocaleMessage(
+                response.data.language_code,
+                response.data.translations
+            )
+            i18n.locale = response.data.language_code
+        },
+        /**
+         * If the user has previously used the app, and set their default
+         * language, then load it from localStorage.
+         */
+        async loadDefaultLanguage(context) {
+            const defaultLanguage = localStorageUtils.getDefaultLanguage()
+            if (defaultLanguage) {
+                context.commit("updateDefaultLanguage", defaultLanguage)
+            }
+        }
```

Comment:
This is currently not working. I don't know what you planned, but you can set `defaultLanguage` in `BaseView.vue`. First you need to add `localStorageUtils.setDefaultLanguage(response.data.default_language_code)` in `fetchTranslations` (set `defaultLanguage` to `localStorage` because it is `null` now) and after that `BaseView.vue` can look like this:
```javascript
<script lang="ts">
import Vue from "vue"
import axios from "axios"
import NavBar from "../components/NavBar.vue"
import SidebarNav from "../components/SidebarNav.vue"
import i18n from "../i18n"
import { TranslationAPIResponse } from "../interfaces"

export default Vue.extend({
    components: {
        NavBar,
        SidebarNav
    },
    computed: {
        username() {
            const user = this.$store.state.user
            return user ? user.username : "Unknown"
        },
        siteName() {
            return this.$store.state.metaModule.siteName
        }
    },
    async mounted() {
        const defaultLanguage = localStorage.getItem(
            "piccoloAdminDefaultLanguage"
        )
        const response = await axios.get<TranslationAPIResponse>(
            `./api/translations/${defaultLanguage}/`
        )
        i18n.setLocaleMessage(
            response.data.language_code,
            response.data.translations
        )
        i18n.locale = response.data.language_code
    }
})
</script>
```
With this we have a `defaultLanguage` persisted even if we refresh the page

Reply:
You're right - it still needs some work. I'll try what you suggested.

Reply:
This also doesn't work well.

Reply:
I haven't had a chance yet.

Reply:
I gave it a go, and I think everything is working now.

Will you try the latest changes?

One thing to be aware of is I changed the URLs slightly. If you're using Nginx locally as a proxy, it needs updating to this:

https://github.com/piccolo-orm/piccolo_admin/blob/7871fd5d43ea21d505d7d7cbf725b168011277b1/extra/piccolo_admin

Now all of the public endpoints are under `/public/` and all of the private endpoints are under `/api/`. It makes things a bit cleaner.